### ISSUE TITLE
"SonarQube Branches & Pull Requests" feature displays configuration logs more than once

### ISSUE THREAD

**Issue by axel3rd** [JetBrains/TeamCity.SonarQubePlugin#82]
Currently, the header configuration log is displayed more than once (one by step in fact):
```
[13:01:16] SonarQube plugin detects pull-request, 'env.SONARQUBE_SCANNER_PARAMS' set with '{"sonar.pullrequest.key":"42","sonar.pullrequest.branch":"Foo bar","sonar.pullrequest.base":"master","sonar.pullrequest.provider":"github","sonar.pullrequest.github.repository":"foo/bar"}'
[13:01:16] SonarQube plugin detects pull-request, 'env.SONARQUBE_SCANNER_PARAMS' set with '{"sonar.pullrequest.key":"42","sonar.pullrequest.branch":"Foo bar","sonar.pullrequest.base":"master","sonar.pullrequest.provider":"github","sonar.pullrequest.github.repository":"foo/bar"}'
[13:01:16] SonarQube plugin detects pull-request, 'env.SONARQUBE_SCANNER_PARAMS' set with '{"sonar.pullrequest.key":"42","sonar.pullrequest.branch":"Foo bar","sonar.pullrequest.base":"master","sonar.pullrequest.provider":"github","sonar.pullrequest.github.repository":"foo/bar"}'
[13:01:16] SonarQube plugin detects pull-request, 'env.SONARQUBE_SCANNER_PARAMS' set with '{"sonar.pullrequest.key":"42","sonar.pullrequest.branch":"Foo bar","sonar.pullrequest.base":"master","sonar.pullrequest.provider":"github","sonar.pullrequest.github.repository":"foo/bar"}'
[13:01:16] SonarQube plugin detects pull-request, 'env.SONARQUBE_SCANNER_PARAMS' set with '{"sonar.pullrequest.key":"42","sonar.pullrequest.branch":"Foo bar","sonar.pullrequest.base":"master","sonar.pullrequest.provider":"github","sonar.pullrequest.github.repository":"foo/bar"}'
[13:01:16] SonarQube plugin detects pull-request, 'env.SONARQUBE_SCANNER_PARAMS' set with '{"sonar.pullrequest.key":"42","sonar.pullrequest.branch":"Foo bar","sonar.pullrequest.base":"master","sonar.pullrequest.provider":"github","sonar.pullrequest.github.repository":"foo/bar"}'
[13:01:16] SonarQube plugin detects pull-request, 'env.SONARQUBE_SCANNER_PARAMS' set with '{"sonar.pullrequest.key":"42","sonar.pullrequest.branch":"Foo bar","sonar.pullrequest.base":"master","sonar.pullrequest.provider":"github","sonar.pullrequest.github.repository":"foo/bar"}'
[13:01:16] SonarQube plugin detects pull-request, 'env.SONARQUBE_SCANNER_PARAMS' set with '{"sonar.pullrequest.key":"42","sonar.pullrequest.branch":"Foo bar","sonar.pullrequest.base":"master","sonar.pullrequest.provider":"github","sonar.pullrequest.github.repository":"foo/bar"}'
```

----

>>    But log "SonarQube plugin detects [pull-request|branch], 'env.SONARQUBE_SCANNER_PARAMS' set with ..." is displayed 3 times ... I should understand why 😒

> Fixed by Execute process once (and not 4 each steps). It was because ParametersPreprocessor is executed for each steps of build.

Initial implementation ([#77](https://github.com/JetBrains/TeamCity.SonarQubePlugin/pull/77#issuecomment-591552880)) has not fixed the problem.

----

https://github.com/JetBrains/TeamCity.SonarQubePlugin/blob/eb0d53dad41183f7be56637a4a8f4863d2fdf3e7/sonar-plugin-server/src/main/java/jetbrains/buildserver/sonarplugin/buildfeatures/BranchesAndPullRequestsParametersPreprocessor.java#L42-L45

The code seems executed at same time => sysenv flag is not suffisant.

Adding a block synchronize is perhaps the solution. @Linfar : Is it tolerable in a `ParametersPreprocessor` ?

**Comment by Linfar** [comment]
Synchronization can only solve problems caused by several threads concurrently modifying a shared state. Here there is only one thread initializing AgentBuild so synchronization won't help. Since runParameters isn't used anyway then probably BranchesAndPullRequestsParametersPreprocessor could be rewritten to BuildStartContextProcessor?

**Comment by axel3rd** [comment]
I'm currently /OOO, but I will try to achieve that this month.

**Comment by axel3rd** [comment]
Historical link: commit [0b78ca0](https://github.com/JetBrains/TeamCity.SonarQubePlugin/commit/0b78ca03da665f48239f52d66d87430ee2123be0)

**Comment by axel3rd** [comment]
Fixed by #84 (merged)
